File: LeetCodeProblemsCMake/Problems/LeetCode1306JumpGameIII.cpp
//
// Created by Saeed Jannati on 5/17/2026 AD.
//

#include "LeetCode1306JumpGameIII.h"

#include <queue>
using namespace std;

bool LeetCode1306JumpGameIII::canReach(vector<int> &arr, int start) {
    if (arr[start] == 0) {
        return true;
    }
    int length = static_cast<int>(arr.size());
    vector<bool> visited(length);
    visited[start] = true;
    queue<int> queue{};
    queue.push(start);
    int next{};
    while (!queue.empty()) {
        int index = queue.front();
        queue.pop();
        for (int i = -1; i < 2; i = i + 2) {
            next=index+i*arr[index];
            if (next<0)
                continue;
            if (next>=length)
                continue;
            if (visited[next])
                continue;
            if (arr[next] == 0)
                return true;
            visited[next] = true;
            queue.push(next);
        }
    }
    return false;
}

File: LeetCodeProblemsJava/src/Problems/LeetCode1306JumpGameIII.java
package Problems;

public class LeetCode1306JumpGameIII {
    void dfs(int[] arr, boolean[] visited, int length, int index, boolean[] reached) {
        if (reached[0])
            return;
        int next = index + arr[index];
        if (next < length && next >= 0) {
            if (!visited[next]) {
                if (arr[next] == 0) {
                    reached[0] = true;
                    return;
                }
                visited[next] = true;
                dfs(arr, visited, length, next, reached);
                visited[next] = false;
            }
        }
        next = index - arr[index];
        if (next < length && next >= 0) {
            if (!visited[next]) {
                if (arr[next] == 0) {
                    reached[0] = true;
                    return;
                }
                visited[next] = true;
                dfs(arr, visited, length, next, reached);
                visited[next] = false;
            }
        }
    }

    public boolean canReach(int[] arr, int start) {
        if (arr[start] == 0)
            return true;
        int length = arr.length;
        boolean[] visited = new boolean[length];
        boolean[] reached = new boolean[1];
        visited[start] = true;
        dfs(arr, visited, length, start, reached);
        return reached[0];
    }
}
